### PR TITLE
feat(rewards): add format-aware reward routing

### LINKED ISSUE #101: [Task 18][QianCyrus] Format-aware reward router
## Task

Relax Contributor Program 2026 Cohort 1 — Task 18: Format-aware reward router.

Claim record: https://github.com/redai-infra/Relax/issues/86#issuecomment-5077510261

## 1. Repository, baseline commit, and development branch

- Repository: `redai-infra/Relax`
- Baseline branch: `main`
- Baseline commit: `039ce876d25540adad847d4223b4de4722d8f425`
- Development branch: `feat/format-aware-reward-router`
- Delivery: GitHub Issue + Draft Pull Request

The implementation and PR will be limited to Task 18. No unrelated formatting, algorithm, scheduler, or performance changes are in scope.

### CPU/unit-test environment

- OS: Red Hat Enterprise Linux 9.4
- Python: 3.12.13
- PyTest: 9.1.1
- PyTorch: 2.11.0+cu130
- CPU architecture: ARM64/aarch64
- GPU: not required by the router unit tests

### End-to-end validation environment

- Environment type: isolated Mamba environment built from scratch; no container image
- OS: Red Hat Enterprise Linux 9.4
- Python: 3.12
- GPU: 1 × NVIDIA GH200, 97,871 MiB visible memory
- Driver: 595.58.03
- CUDA toolkit: 13.2
- PyTorch: 2.11.0+cu130
- Ray: 2.56.1
- SGLang: 0.5.12.post1
- Transformer Engine: 2.14.1
- cuDNN: 9.19
- cuDNN frontend: 1.26
- Host architecture: ARM64/aarch64
- Parallelism: single-node, single-GPU colocate

The exact CPU model and host RAM were not part of the E2E acceptance snapshot and will not be used as a correctness metric. GPU identity, software versions, source commit, data hash, command, step count, and raw training log will be recorded.

### Targeted tests

```bash
python -m pytest -v \
  tests/engine/rewards/test_reward_router.py \
  tests/engine/rewards/test_reward_worker.py \
  tests/utils/test_arguments_opd_teacher_colocate.py
```

### Reward regression tests

```bash
python -m pytest -v \
  tests/engine/rewards/ \
  tests/utils/test_arguments_opd_teacher_colocate.py
```

### Repository checks

```bash
pre-commit run --all-files
```

### Mixed-reward 10-update E2E configuration

```bash
export RM_TYPE=
export RM_TYPE_FALLBACK=error
export NUM_ROLLOUT=10
export ROLLOUT_BATCH_SIZE=4
export N_SAMPLES=4
export GLOBAL_BATCH_SIZE=16
export ROLLOUT_MAX_RESPONSE_LEN=4096
export MAX_TOKENS_PER_GPU=4480
export LOGPROBS_MAX_TOKENS_PER_GPU=4480
export SGLANG_MEM_FRACTION_STATIC=0.35
export ATTENTION_BACKEND=auto

bash examples/beginner-task/run-qwen3-0.6B-1xgpu-grpo-mixed-reward.sh
```

Update arithmetic:

```text
10 rollouts × 4 prompts × 4 samples / global batch 16
= 10 optimizer updates
```

The E2E dataset contains 20 GSM8K prompts and 20 MMLU prompts. Each shuffled rollout batch contains two math prompts and two multiple-choice prompts.

### Planned files

- `relax/engine/rewards/registry.py`
- `relax/engine/rewards/__init__.py`
- `relax/utils/arguments.py`
- `tests/engine/rewards/test_reward_router.py`
- `tests/engine/rewards/test_reward_worker.py`
- `tests/utils/test_arguments_opd_teacher_colocate.py`
- `docs/en/guide/customize-training.md`
- `docs/zh/guide/customize-training.md`

### Planned design

1. Add a `RewardSpec` containing the reward name, sync/async mode, handler, and optional label matcher.
2. Add a `RewardRegistry` as the single source of reward registrations.
3. Add a pure `resolve_reward_route()` function that returns a structured route decision without Ray or GPU dependencies.
4. Preserve explicit `--rm-type` semantics for backward compatibility.
5. Add `--rm-type-fallback` with three policies:
   - `zero`: return zero and emit a warning;
   - `error`: emit a warning and raise a descriptive error;
   - a registered reward name: dispatch to that reward.
6. Detect unknown, missing, ambiguous, and metadata/label-conflicting routes.
7. Dispatch registered synchronous rewards through the existing worker pool and await registered asynchronous rewards directly.
8. Keep `boxed_` behavior compatible.
9. Deduplicate repeated warnings by route failure identity.

Routing precedence:

```text
metadata.rm_type
  → explicit --rm-type compatibility path
  → exactly one registered label matcher
  → configured --rm-type-fallback
```

The explicit CLI route remains ahead of automatic label inference so an existing single-task command keeps its previous behavior.

### Non-goals

This task will not:

- change math or multiple-choice scoring formulas;
- modify GRPO/PPO/DAPO algorithms;
- modify Controller, rollout scheduling, or weight synchronization;
- add third-party dependencies;
- change public `Sample.label` or `Sample.metadata` structures;
- introduce a general natural-language task classifier;
- modify existing datasets in bulk;
- claim performance or convergence improvements.

## 5. Baseline data, behavior, metrics, and targets

At baseline commit `039ce876d25540adad847d4223b4de4722d8f425`:

- `metadata.rm_type` can override the global `--rm-type`;
- dispatch is duplicated across a worker `if/elif` chain and async/sync type collections;
- no unified registry exists;
- no label-based route inference exists;
- no configurable fallback policy exists;
- no metadata/label conflict detection exists;
- an unknown type reaches a Ray worker and fails late with `RewardWorker: unknown rm_type=...`;
- a missing type fails with `Rule-based RM type is not specified.`;
- a valid but incorrect metadata type can silently invoke the wrong scorer.

Acceptance targets:

| Requirement | Measurement | Target |
|---|---|---|
| Mixed batch | Math + multiple-choice samples in one batch | Every sample reaches the expected reward |
| Metadata route | Valid `metadata.rm_type` | Dispatches to that registered reward |
| CLI compatibility | Existing explicit `--rm-type` | Existing route remains unchanged |
| Label route | No metadata or CLI, exactly one matcher | Dispatches to the matching reward |
| Conflict handling | Metadata disagrees with a recognized label | Configured fallback with `conflicting_reward_types` warning |
| Unknown/missing handling | Invalid or absent type | Configured zero/error/reward fallback |
| Registry extensibility | Register a test reward | One registration, no generic router branch change |
| Sync/async handling | Registered handlers of both modes | Correct execution path |
| Unit tests | Reward router/worker/argument tests | All pass |
| Code checks | `pre-commit run --all-files` | All pass |
| E2E integration | 10 rollouts and 10 optimizer updates | Complete, finite, no fallback/OOM/Traceback |

### Unit-level verification

- registry lookup and deterministic ordering;
- duplicate, empty, and invalid registrations;
- conservative math and multiple-choice label matchers;
- metadata, CLI, and label priority;
- mixed math + multiple-choice batch;
- unknown, invalid, missing, ambiguous, and conflicting routes;
- `zero`, `error`, and registered reward fallbacks;
- sync/async dispatch;
- `boxed_` compatibility;
- warning deduplication;
- legacy explicit `--rm-type` behavior.

### End-to-end verification

Public datasets:

- `openai/gsm8k`, revision `740312add88f781978c0658806c59bc2815b9866`, first 20 training rows;
- `cais/mmlu`, revision `c30699e8356da336a370243923dbaf21066bb9fe`, first 20 `high_school_computer_science` test rows.

Prepared JSONL SHA256:

```text
1a5a7bcf836a2e35f1f208ba741929464fb10b0c19fbc4bd5df3bfef6c78efe0
```

The E2E run will use no global `--rm-type` and `--rm-type-fallback error`. Therefore a missing, unknown, or conflicting sample route fails the run instead of being hidden by a zero reward.

Expected E2E counts:

```text
40 prompts
160 generated responses
80 math reward evaluations
80 multiple-choice reward evaluations
10 completed rollouts
10 unique optimizer updates, steps 0..9
```

E2E acceptance also requires finite policy loss, KL, and gradient norm; no OOM, mixed CUDA runtime, actor failure, or traceback; and a final validator bound to the unique run identity.

## 7. Expected delivery

The Draft PR will include:

- the reward registry and pure router;
- metadata/CLI/label routing;
- configurable fallback behavior and warnings;
- sync/async registry dispatch;
- backward compatibility tests;
- mixed-batch, conflict, fallback, and registry tests;
- English and Chinese documentation;
- reproducible test commands;
- a sanitized 10-step E2E result summary and relevant log excerpts.

### PR DESCRIPTION
**Pull request by QianCyrus**
## What / Summary

- Closes #101.
- Adds a format-aware reward registry and per-sample router for mixed-task batches.
- Routes samples from `metadata.rm_type`, the backward-compatible global `--rm-type`, or conservative label matchers.
- Adds configurable handling for unknown, missing, ambiguous, and conflicting routes.
- Keeps existing reward scoring formulas and worker isolation unchanged.

## Why

The baseline can use `metadata.rm_type` as a per-sample override, but reward selection is spread across a worker `if/elif` chain and separate sync/async type collections. It has no unified label-matcher registry, configurable fallback policy, or metadata/label conflict detection.

Consequences include:

- adding a reward requires editing multiple dispatch locations;
- unknown reward types fail late inside a Ray worker;
- missing types fail with a generic error;
- valid but incorrect metadata can invoke the wrong scorer without detecting a conflict;
- mixed datasets without complete metadata cannot infer math versus multiple-choice routes.

### Registry and pure router

- Adds `RewardSpec`, `RewardRegistry`, and `RouteDecision` in `relax/engine/rewards/registry.py`.
- Adds pure `resolve_reward_route()` logic with no Ray or GPU dependency.
- Adds conservative label matchers for:
  - numeric, fractional, and explicit LaTeX math labels;
  - answer-tagged single-letter multiple-choice labels and explicit choices/options metadata.

Routing precedence:

```text
metadata.rm_type
  → explicit --rm-type compatibility path
  → exactly one registered label matcher
  → configured --rm-type-fallback
```

The explicit CLI route remains ahead of automatic inference so existing single-task commands retain their previous behavior.

### Unified execution

- Replaces reward-specific `RewardWorker.compute()` branches with registry lookup.
- Dispatches registered synchronous handlers through the existing Ray worker pool.
- Awaits registered asynchronous handlers directly.
- Preserves `boxed_` response extraction behavior.
- Centralizes built-in reward registrations; adding a reward requires a handler adapter and one registration rather than a generic router branch.

### Fallback policy

Adds:

```text
--rm-type-fallback zero
--rm-type-fallback error
--rm-type-fallback <registered-reward>
```

- `zero`: emits a warning and returns a zero reward;
- `error`: emits a warning and raises a descriptive `ValueError`;
- registered reward: emits a warning and dispatches to that fallback.

Warnings contain the reason, source, requested type, action, selected type, group index, and sample index. Repeated equivalent route failures are logged once per executor.

Structured failure reasons include:

- `invalid_metadata_type`
- `unknown_metadata_type`
- `conflicting_reward_types`
- `invalid_cli_type`
- `unknown_cli_type`
- `ambiguous_label_type`
- `missing_reward_type`

### Documentation and tests

- Documents routing precedence, fallback behavior, and one-line registry extension in English and Chinese.
- Adds registry, matcher, priority, fallback, mixed-batch, conflict, warning, worker, and argument tests.

### Relevant unit and regression tests

Environment:

```text
Red Hat Enterprise Linux 9.4
Python 3.12.13
PyTest 9.1.1
PyTorch 2.11.0+cu130
```

Command:

```bash
python -m pytest -q \
  tests/engine/rewards/ \
  tests/utils/test_arguments_opd_teacher_colocate.py
```

Result:

```text
49 passed
```

### Repository checks

Command:

```bash
CC=gcc CXX=g++ CGO_ENABLED=0 pre-commit run --all-files
```

Result:

```text
check for added large files     Passed
check for merge conflicts       Passed
check for broken symlinks       Passed
detect private key              Passed
fix end of files                Passed
debug statements                Passed
ruff                            Passed
ruff format                     Passed
mdformat                        Passed
clang-format                    Passed
docformatter                    Passed
check-conflict-markers          Passed
format-ascii-boxes              Passed
Detect hardcoded secrets        Passed
```

The explicit compiler settings avoid a cluster login-node `nvc` incompatibility while installing the Go-based gitleaks hook. The hooks themselves all pass, and the checkout remains clean.

### 10-update end-to-end mixed-reward GRPO

This additional integration run used public data, a real model, and one GPU; it was not a synthetic unit-test-only run.

Environment:

```text
1 × NVIDIA GH200, 97,871 MiB visible memory
CUDA toolkit 13.2
PyTorch 2.11.0+cu130
Ray 2.56.1
SGLang 0.5.12.post1
Transformer Engine 2.14.1
Qwen3-0.6B
single-GPU colocate
```

Data:

- `openai/gsm8k`, revision `740312add88f781978c0658806c59bc2815b9866`, first 20 training rows;
- `cais/mmlu`, revision `c30699e8356da336a370243923dbaf21066bb9fe`, first 20 `high_school_computer_science` test rows;
- prepared JSONL SHA256: `1a5a7bcf836a2e35f1f208ba741929464fb10b0c19fbc4bd5df3bfef6c78efe0`.

Configuration:

```text
global --rm-type: unset
--rm-type-fallback: error
rollouts: 10
prompts per rollout: 4
samples per prompt: 4
global batch: 16
optimizer updates: 10 × 4 × 4 / 16 = 10
response cap: 4096
actor/log-prob token budget: 4480
SGLang static memory fraction: 0.35
attention backend: auto / Transformer Engine FusedAttention
```

Result:

| Check | Observed |
|---|---:|
| Completed rollouts | 10/10 |
| Unique optimizer steps | 0..9, 10/10 |
| Generated responses | 160 |
| Math reward evaluations | 80 |
| Multiple-choice reward evaluations | 80 |
| Reward routing fallbacks | 0 |
| OOM / NaN / Inf / Traceback | 0 |
| Mean raw reward | 0.66875 |
| Positive binary rewards | 107/160 |
| Mean policy loss | 0.024132 |
| Mean PPO KL | 0.000619 |
| Maximum gradient norm | 1.3236 |
| Maximum observed total sequence | 4363 tokens |
| Wall time including initialization | 9m 02s |

Per-step metrics:

| Step | Raw reward | Policy loss | PPO KL | Grad norm |
|---:|---:|---:|---:|---:|
| 0 | 1.0000 | 0.000000 | -0.000399 | 0.0000 |
| 1 | 0.6875 | -0.049346 | 0.000723 | 0.6227 |
| 2 | 0.5000 | 0.025843 | 0.000540 | 1.1645 |
| 3 | 0.5625 | 0.024774 | 0.000630 | 1.3236 |
| 4 | 0.4375 | -0.040699 | 0.000314 | 0.6256 |
| 5 | 0.6250 | 0.062120 | 0.000891 | 1.1688 |
| 6 | 0.5000 | 0.070711 | 0.000885 | 0.7640 |
| 7 | 0.7500 | 0.000000 | 0.001212 | 0.0000 |
| 8 | 0.9375 | 0.177224 | 0.000468 | 1.1111 |
| 9 | 0.6875 | -0.029303 | 0.000930 | 0.5852 |

Steps 0 and 7 have finite zero policy loss/gradient norm because their GRPO groups have uniform rewards and therefore zero normalized advantages.

Sanitized final log gate:

```text
RELAX_ACCEPTANCE_COMPLETE expected_updates=10 expected_rollouts=10 job_id=<redacted> run_dir=<redacted>
```

The full raw log contains cluster-local paths and identifiers, so it is not attached publicly. The table above is generated from the validated raw log; additional sanitized excerpts can be provided during review.

### Risks

- Label inference is intentionally conservative; unsupported or ambiguous labels use the configured fallback instead of guessing.
- The default missing-route behavior changes from a generic exception to warning + zero because `--rm-type-fallback` defaults to `zero`.
- Import-time registry contents are treated as read-only during reward execution.
- Metadata/label conflicts for rewards with a matcher now surface instead of silently using the metadata scorer.

### Compatibility

- Valid `metadata.rm_type` remains the highest-priority route.
- Explicit `--rm-type` remains supported and takes precedence over automatic label inference.
- Existing reward scoring functions and return values are preserved.
- `boxed_` types remain supported.
- No new dependencies are added.

### Rollback

- Revert this PR to restore the original worker dispatch and missing-route behavior.
- Existing single-task commands can continue to pass explicit `--rm-type`; no dataset migration is required.

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [x] Documentation update
- [x] Refactoring
- [ ] Performance improvement
- [ ] CI/CD or build changes

## Screenshots / Logs

No screenshot is included because routing correctness is more directly demonstrated by deterministic unit tests, explicit route counts, and the strict E2E log gate. A sanitized log excerpt and the complete 10-step metric table are included above.

## Checklist

- [x] Diff only contains Task 18 changes.
- [x] New and related tests pass.
- [x] `pre-commit run --all-files` passes.
- [x] Documentation and CLI help are updated.
- [x] Mixed math + multiple-choice routing is tested.
- [x] Fallback, conflict, registry, and backward compatibility are tested.
- [x] A real 10-update E2E integration run passes.
- [x] No secrets, complete datasets, checkpoints, internal addresses, or raw private logs are included.
- [x] PR is opened as Draft for design and behavior review.

### CODE REVIEW

**Comment by SigureMo** [comment]
#137 merged，本 PR close～
